### PR TITLE
Add KuduStorage compatibility WIP

### PR DESCRIPTION
**Pull request by samkuz**
This commit will add initial KuduStorage compatibility, as well as initial Kudu integration testing.

### CODE REVIEW

File: src/main/scala/io/phdata/retirementage/SparkDriver.scala [hunk 1]
```diff
@@ -67,8 +72,8 @@ object SparkDriver extends LazyLogging {
   def getFilter(database: Database, table: DatedTable) = {
     table.storage_type match {
       case "parquet" => new DatedTableFilter(database, table) with HdfsStorage
```
Comment (afoerster) [inline src/main/scala/io/phdata/retirementage/SparkDriver.scala:69]: Where's the "avro" case?

File: src/main/scala/io/phdata/retirementage/filters/KuduTableFilter.scala [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    with StorageActions {
+
+  // New currentFrame for KuduTables
+  override lazy val currentFrame =
+    spark.sqlContext.read
+      .options(
+        Map("kudu.master" -> "localhost:7051", "kudu.table" -> qualifiedTableName))
+      .kudu
+      .cache()
+  // taken from tablefilter
```
Comment (afoerster) [inline src/main/scala/io/phdata/retirementage/filters/KuduTableFilter.scala]: no need for this comment

File: src/main/scala/io/phdata/retirementage/filters/KuduTableFilter.scala [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+abstract class KuduTableFilter(database: Database, table: DatedTable)
+    extends TableFilter(database, table)
+    with StorageActions {
+
+  // New currentFrame for KuduTables
+  override lazy val currentFrame =
+    spark.sqlContext.read
+      .options(
+        Map("kudu.master" -> "localhost:7051", "kudu.table" -> qualifiedTableName))
```
Comment (afoerster) [inline src/main/scala/io/phdata/retirementage/filters/KuduTableFilter.scala]: These hardcoded values need to be passed in

File: src/main/scala/io/phdata/retirementage/filters/KuduTableFilter.scala [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      .kudu
+      .cache()
+  // taken from tablefilter
+  lazy val dateExpression: Column =
+    getDateExpression(currentFrame, table.expiration_column, table.date_format_string)
+
+  val DEFAULT_STRING_FORMAT = "yyyy-MM-dd"
```
Comment (afoerster) [inline src/main/scala/io/phdata/retirementage/filters/KuduTableFilter.scala]: shouldn't need to redefine this

File: src/main/scala/io/phdata/retirementage/filters/KuduTableFilter.scala [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      minDate.filter(date_add(col(table.expiration_column), table.expiration_days) < current_date())
+
+    !minDateFrame.rdd
+      .isEmpty()
+  }
+
+  def filteredFrame() = expiredRecords()
+
+  // Taken from DatedTableFilter
+  def getDateExpression(df: DataFrame, dateColumn: String, stringFormat: Option[String]): Column = {
```
Comment (afoerster) [inline src/main/scala/io/phdata/retirementage/filters/KuduTableFilter.scala]: Shoudln't need to redefine/copy this

File: src/main/scala/io/phdata/retirementage/SparkDriver.scala [hunk 6]
```diff
@@ -52,7 +57,11 @@ object SparkDriver extends LazyLogging {
             filter.noAction()
           } else {
             // table has no hold and needs to be filtered, write a new dataset without expired records
-            filter.doFilter(computeCountsFlag, dryRun = false)
+            if (filter.isInstanceOf[KuduTableFilter]) {
+              filter.doKuduFilter(computeCountsFlag, dryRun = false)
```
Comment (afoerster) [inline src/main/scala/io/phdata/retirementage/SparkDriver.scala]: there should just be 'filter' method and the kudu implementation should override 'filter'

File: src/main/scala/io/phdata/retirementage/storage/KuduStorage.scala [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+trait KuduStorage extends StorageActions with LazyLogging {
+  override def persistFrame(computeCountsFlag: Boolean,
+                            dryRun: Boolean,
+                            qualifiedTableName: String,
+                            storageType: String,
+                            currentFrame: DataFrame,
+                            filteredFrame: DataFrame): RetirementReport = {
+    try {
+      val kuduContext =
```
Comment (afoerster) [inline src/main/scala/io/phdata/retirementage/storage/KuduStorage.scala]: A lot (or all) of this logic is duplicated and needs to be extracted

File: src/main/scala/io/phdata/retirementage/storage/KuduStorage.scala [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+                         false,
+                         DatasetReport(getCurrentDatasetLocation(qualifiedTableName)),
+                         None,
+                         Some(e.getMessage))
+      }
+    }
+  }
+
+  def getNewDatasetLocation(qualifiedTableName: String): String = {
+    "nokudu"
```
Comment (afoerster) [inline src/main/scala/io/phdata/retirementage/storage/KuduStorage.scala]: would it make sense to just return the tablename here?

**Review by afoerster — CHANGES_REQUESTED** [review]

File: src/it/scala/io/phdata/retirementage/KuduStorageTest.scala [hunk 9]
```diff
@@ -0,0 +1,218 @@
...
+
+import scala.collection.JavaConverters._
+
+class KuduStorageTest extends FunSuite with SparkTestBase {
+  val kuduMaster    = "localhost:7051"
+  val kuduContext   = new KuduContext(kuduMaster, spark.sqlContext.sparkContext)
+  val defaultSchema = StructType(StructField("date", LongType, false) :: Nil)
+  val defaultKey    = Seq("date")
+
+  test("creating a kudu table") {
```
Comment (afoerster) [inline src/it/scala/io/phdata/retirementage/KuduStorageTest.scala]: Tests that aren't essential to the application can be removed (this was good for figuring out how to create a table, but doesn't help test the application)

File: src/it/scala/io/phdata/retirementage/KuduStorageTest.scala [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+    val kuduTableName = createKuduTable(defaultSchema, defaultKey, kuduContext)
+    println("Created kuduTableName: " + kuduTableName)
+    val table              = DatedTable(kuduTableName, "kudu", "date", 1, None, None, None)
+    val database           = Database("default", Seq(table))
+    val qualifiedTableName = s"${database.name}.${kuduTableName}"
+
+    assert(kuduContext.tableExists(qualifiedTableName))
+  }
+
+  test("insert rows") {
```
Comment (afoerster) [inline src/it/scala/io/phdata/retirementage/KuduStorageTest.scala]: This test can be removed

File: src/it/scala/io/phdata/retirementage/KuduStorageTest.scala [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+
+    val testDf =
+      spark.sqlContext.read
+        .options(Map("kudu.master" -> kuduMaster, "kudu.table" -> qualifiedTableName))
+        .kudu
+
+    assertResult(3)(testDf.count())
+  }
+
+  test("delete rows") {
```
Comment (afoerster) [inline src/it/scala/io/phdata/retirementage/KuduStorageTest.scala]: Can you add a more descriptive title?

File: src/it/scala/resources/hive-site.xml [hunk 12]
```diff
@@ -0,0 +1,17 @@
+<configuration>
```
Comment (afoerster) [inline src/it/scala/resources/hive-site.xml]: No sure we need this file? Where was this copied from?

**Review by afoerster — CHANGES_REQUESTED** [review]

File: src/it/scala/resources/hive-site.xml (hunk 12 above)
Comment (samkuz) [reply src/it/scala/resources/hive-site.xml]: I just copied the resources from test, I wasn't sure if they were important I'll delete it

**Comment by afoerster** [comment]
@samkuz have you been able to test this on the cluster?

File: README.md [hunk 13]
```diff
@@ -150,10 +150,11 @@ application will filter out all records that are past their expiration date. You
 this similarly to how you would filter out a parquet, or avro table.
 
 ## Configuration File
-It is necessary to specify at least one Kudu Master in the configuration file. it is not necessary
+It is necessary to specify at least one Kudu Master in the configuration file. It is not necessary
 to specify a database for a Kudu table. If you created the Kudu table in Impala you will have to
 specify the database (ex. impala::database1.table1), but if the Kudu table is not connected to any
-database then you would not define a database.
+database then you would not define a database. In order to not define a database you would input
```
Comment (afoerster) [inline README.md]: this extra space before the line will mix up the formatting

File: src/main/scala/io/phdata/retirementage/filters/TableFilter.scala [hunk 14]
```diff
@@ -34,7 +34,18 @@ abstract class TableFilter(database: Database, table: Table)
...
 
   lazy val currentFrame: DataFrame = getCurrentFrame(qualifiedTableName)
-  val qualifiedTableName           = s"${database.name}.${table.name}"
+  val qualifiedTableName           = setQualifiedName(database, table)
+
+  /**
+    * Returns the qualified table name for a given database and table
+    */
+  def setQualifiedName(database: Database, table: Table): String = {
+    if (database.name.equals("")) {
```
Comment (afoerster) [inline src/main/scala/io/phdata/retirementage/filters/TableFilter.scala]: Can you add a comment here which conditions the database name would be empty?

**Review by afoerster — CHANGES_REQUESTED** [review]
Small fixes

**Comment by afoerster** [comment]
LGTM

**Review by afoerster — APPROVED** [review]

**Pull request merged**